frontends/web/src/routes/account/send/send.tsx

```diff
@@ -270,44 +278,70 @@ class Send extends Component<Props, State> {
             return;
         }
         const txInput = this.txInput();
-        apiPost('account/' + this.getAccount()!.code + '/tx-proposal', txInput).then(result => {
-            this.setState({ valid: result.success });
-            if (result.success) {
-                this.setState({
-                    addressError: undefined,
-                    amountError: undefined,
-                    dataError: undefined,
-                    proposedFee: result.fee,
-                    proposedAmount: result.amount,
-                    proposedTotal: result.total,
-                });
-                if (updateFiat) {
-                    this.convertToFiat(result.amount.amount);
-                }
-            } else {
-                const errorCode = result.errorCode;
-                switch (errorCode) {
-                    case 'invalidAddress':
-                        this.setState({ addressError: this.props.t('send.error.invalidAddress') });
-                        break;
-                    case 'invalidAmount':
-                    case 'insufficientFunds':
-                        this.setState({ amountError: this.props.t(`send.error.${errorCode}`) });
-                        break;
-                    case 'invalidData':
-                        this.setState({ dataError: this.props.t(`send.error.invalidData`) });
-                        break;
-                    default:
-                        this.setState({ proposedFee: undefined });
-                        if (errorCode) {
-                            this.unregisterEvents();
-                            alertUser(errorCode, this.registerEvents);
-                        }
+        if (this.proposeTimeout) {
+            clearTimeout(this.proposeTimeout);
+        }
+        this.setState({ isUpdatingProposal: true });
+        this.proposeTimeout = setTimeout(() => {
+            const propose = apiPost('account/' + this.getAccount()!.code + '/tx-proposal', txInput)
+            .then(result => {
+                const pos = this.pendingProposals.indexOf(propose);
+                if (this.pendingProposals.length - 1 === pos) {
+                    this.txProposal(updateFiat, result);
                 }
+                this.pendingProposals.splice(pos, 1);
+            })
+            .catch(() => {
+                this.setState({ valid: false });
+                this.pendingProposals.splice(this.pendingProposals.indexOf(propose), 1);
+            });
+            this.pendingProposals.push(propose);
```

Comment:
confused about js semantics. could the `then()` clause happen before this push is ever executed?

Reply:
No it could not happen first, even if there was no fetch at all
```
Promise.resolve()
  .then(() => console.log('then'));
console.log('I am first')
```